mater/lib/src/file_reader.rs

```diff
@@ -0,0 +1,246 @@
+use std::{
+    collections::{HashMap, VecDeque},
+    io::Cursor,
+    ops::Deref,
+    path::Path,
+};
+
+use async_stream::try_stream;
+use futures::{Stream, TryStreamExt};
+use ipld_core::{cid::Cid, codec::Codec};
+use ipld_dagpb::{DagPbCodec, PbNode};
+use tokio::{
+    fs::File,
+    io::{AsyncRead, AsyncSeek, AsyncSeekExt, AsyncWriteExt},
+};
+
+use crate::{multicodec, v1::BlockMetadata, v2, Error};
+
+/// CAR file loader.
+pub struct FileLoader<R>
+where
+    R: AsyncRead + AsyncSeek + Unpin,
+{
+    reader: v2::Reader<R>,
+    index: HashMap<Cid, PartialNode>,
+}
+
+impl FileLoader<File> {
+    /// Creates a [`FileLoader`] from the given file path.
+    pub async fn from_path<P>(path: P) -> Result<Self, Error>
+    where
+        P: AsRef<Path>,
+    {
+        let file = File::open(path).await?;
+        let mut loader = Self {
+            reader: v2::Reader::new(file),
+            index: HashMap::new(),
+        };
+        loader.index().await?;
+        Ok(loader)
+    }
+}
+
+impl FileLoader<Cursor<Vec<u8>>> {
+    /// Creates a [`FileLoader`] from a vector of bytes.
+    pub async fn from_vec(vec: Vec<u8>) -> Result<Self, Error> {
+        let mut loader = Self {
+            reader: v2::Reader::new(Cursor::new(vec)),
+            index: HashMap::new(),
+        };
+        loader.index().await?;
```

Comment:
What if the file is already indexed?

Reply:
Re-indexed, there's no caching + it's naive (meaning it ignores the file index) and this is the constructor after all, so it assumes "nothing was here before"